### DISCUSSION TITLE
Support MERGE statement

### DISCUSSION THREAD

**Discussion by h-vetinari** [sqlalchemy/sqlalchemy#10603] [Blue Sky]
### Describe the use case

Since postgres 15, there's support for the [`MERGE`](https://www.postgresql.org/docs/current/sql-merge.html) statement, which is basically a better upsert.

Given that the extensive sqlalchemy [docs](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html) on postgres describe `INSERT ... ON CONFLICT` in [detail](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#insert-on-conflict-upsert), but do not mention `MERGE` at all, it is my suspicion that this might not be supported yet, especially as postgres 15 is relatively recent.

PS. I've tried to search the issue tracker, but I couldn't find a mention of this.

### Databases / Backends / Drivers targeted

Postgres

### Example Use

I don't know sqlalchemy well enough to propose what the API should look like. The below is like a kid's crayon version of what things might look like.

```python
>>> from sqlalchemy.dialects.postgresql import insert
>>> merge_stmt = insert(my_table).merge(other_table, join_predicate)\
...     .when_matched(...)\
...     .when_not_matched(...)\
>>> print(merge_stmt)
MERGE INTO my_table
USING other_table
ON join_predicate
WHEN MATCHED THEN
  ...
WHEN NOT MATCHED THEN
  ...;
```

Some more example queries for this features can be found [here](https://www.crunchydata.com/blog/a-look-at-postgres-15-merge-command-with-examples).

**Comment by CaselIT** [comment]
Hi,

currently merge is not supported in any backend.
It would also not be initiated from insert since merge can do other things other than inserting rows.

There are currently no plan on supporting it, meaning that's not planned for a near release of SQLalchemy

**Comment by h-vetinari** [comment]
Thanks for the quick response! I get that it's complicated to design a whole new API around that, and ensure it fits well with everything else (though not surprised that my crude sketch was wrong 😅).

**Comment by zzzeek** [comment]
this is a dupe of #960 and/or #8321 so if you have plans to design and implement this API we can keep it with the original.  

the two variants of "MERGE" here are:

1. we use MERGE to implement "ON CONFLICT" style functionality for Oracle, SQL Server, others and put it behind a generic facade #8321 
2. we implement the full blown SQL MERGE construct note for note and have it work for those backends that support it; this still does not supply a cross-platform "ON CONFLICT" construct 

IMO item 1 is vastly more useful and nobody really needs a SQL standard formalized version of MERGE
